nighthawk/source/common/sequencer_impl.cc

```diff
@@ -0,0 +1,113 @@
+#include "nighthawk/common/exception.h"
+
+#include "common/common/assert.h"
+
+#include "nighthawk/source/common/sequencer_impl.h"
+
+using namespace std::chrono_literals;
+
+namespace Nighthawk {
+
+const std::chrono::milliseconds SequencerImpl::EnvoyTimerMinResolution = 1ms;
+
+SequencerImpl::SequencerImpl(Envoy::Event::Dispatcher& dispatcher, Envoy::TimeSource& time_source,
+                             RateLimiter& rate_limiter, SequencerTarget& target,
+                             std::chrono::microseconds duration,
+                             std::chrono::microseconds grace_timeout)
+    : Sequencer(target), dispatcher_(dispatcher), time_source_(time_source),
+      rate_limiter_(rate_limiter), duration_(duration), grace_timeout_(grace_timeout),
+      start_(time_source.monotonicTime().min()), targets_initiated_(0), targets_completed_(0),
+      spin_when_idle_(true) {
+  if (target_ == nullptr) {
+    throw NighthawkException("SequencerImpl must be constructed with a SequencerTarget.");
+  }
+  periodic_timer_ = dispatcher_.createTimer([this]() { run(true); });
+  incidental_timer_ = dispatcher_.createTimer([this]() { run(false); });
+}
+
+void SequencerImpl::start() {
+  start_ = time_source_.monotonicTime();
+  run(true);
+}
+
+void SequencerImpl::scheduleRun() { periodic_timer_->enableTimer(EnvoyTimerMinResolution); }
+
+void SequencerImpl::stop() {
+  periodic_timer_->disableTimer();
+  incidental_timer_->disableTimer();
+  periodic_timer_.reset(nullptr);
+  incidental_timer_.reset(nullptr);
+  dispatcher_.exit();
+}
+
+void SequencerImpl::run(bool from_timer) {
+  const auto now = time_source_.monotonicTime();
+
+  if ((now - start_) > (duration_)) {
```

Comment:
Can you add some comments explaining the flow of this method? E.g. tell the narrative "Now we check for X, then we check for Y then we do Z..".